audit-service/src/contract-tests/java/uk/gov/gchq/palisade/contract/audit/kafka/KafkaContractTest.java

```diff
@@ -0,0 +1,299 @@
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyString;
+
+/**
+ * An external requirement of the service is to connect to a pair of upstream kafka topics.
+ * <ol>
+ *     <li>The "error" topic can be written to by any service that encounters an error when processing a request</li>
+ *     <li>The "success" topic should only be written to by the filtered-resource-service or the data-service</li>
+ * </ol>
+ * This service does not write to a downstream topic
+ */
+@SpringBootTest(
+        classes = AuditApplication.class,
+        webEnvironment = WebEnvironment.RANDOM_PORT,
+        properties = {"akka.discovery.config.services.kafka.from-config=false"}
+)
+@Import(KafkaContractTest.KafkaInitializer.Config.class)
+@ContextConfiguration(initializers = {KafkaContractTest.KafkaInitializer.class})
+@ActiveProfiles({"akkatest"})
+class KafkaContractTest {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(KafkaContractTest.class);
+    private static final ObjectMapper MAPPER = new ObjectMapper();
+
+    // Serialiser for upstream test input
+    static class RequestSerializer implements Serializer<JsonNode> {
+        @Override
+        public byte[] serialize(final String s, final JsonNode auditRequest) {
+            try {
+                return MAPPER.writeValueAsBytes(auditRequest);
+            } catch (JsonProcessingException e) {
+                throw new SerializationFailedException("Failed to serialize " + auditRequest.toString(), e);
+            }
+        }
+    }
+
+    @Autowired
+    private TestRestTemplate restTemplate;
+    @Autowired
+    private ActorSystem akkaActorSystem;
+    @Autowired
+    private Materializer akkaMaterializer;
+    @Autowired
+    private ConsumerTopicConfiguration consumerTopicConfiguration;
+    @SpyBean
+    private AuditService auditService;
+
+    @Test
+    @DirtiesContext
+    void testErrorRequestSet() {
+        // Add some messages on the error topic
+        // The ContractTestData.REQUEST_TOKEN maps to partition 0 of [0, 1, 2], so the akkatest yaml connects the consumer to only partition 0
+        final Stream<ProducerRecord<String, JsonNode>> requests = ContractTestData.ERROR_RECORD_NODE_FACTORY.get().limit(3L);
+
+        // When - we write to the input
+        ProducerSettings<String, JsonNode> producerSettings = ProducerSettings
+                .create(akkaActorSystem, new StringSerializer(), new RequestSerializer())
+                .withBootstrapServers(KafkaInitializer.kafka.getBootstrapServers());
+
+        Source.fromJavaStream(() -> requests)
+                .runWith(Producer.plainSink(producerSettings), akkaMaterializer)
+                .toCompletableFuture().join();
+
+        // Then - check the audit service has invoked the audit method 3 times
+        Mockito.verify(auditService, Mockito.timeout(3000).times(3)).audit(anyString(), any());
```

Comment:
I may be missing something here, but I does not look like it is testing if the messages are of AuditErrorMessage or AuditSucccessMessage. Only that there auditservice method audit get called 3 times.

Reply:
We do not need to test the message type as they are being picked up by different runnable graphs for the different input topics. We need to make sure that the audit service is invoked though.